Don't suggest restricting bound with unstable traits on stable and mention it's unstable on nightly

estebank opened the pull request:
On nightly, we mention the trait is unstable

```
error[E0277]: the trait bound `T: Unstable` is not satisfied
  --> $DIR/unstable-trait-suggestion.rs:13:9
   |
LL |     foo(t)
   |     --- ^ the trait `Unstable` is not implemented for `T`
   |     |
   |     required by a bound introduced by this call
   |
note: required by a bound in `foo`
  --> $DIR/unstable-trait-suggestion.rs:9:11
   |
LL | fn foo<T: Unstable>(_: T) {}
   |           ^^^^^^^^ required by this bound in `foo`
help: consider restricting type parameter `T` but it is an `unstable` trait
   |
LL | pub fn demo<T: Unstable>(t: T) {
   |              ++++++++++
```

On stable, we don't suggest the trait at all

```
error[E0277]: the trait bound `T: Unstable` is not satisfied
  --> $DIR/unstable-trait-suggestion.rs:13:9
   |
LL |     foo(t)
   |     --- ^ the trait `Unstable` is not implemented for `T`
   |     |
   |     required by a bound introduced by this call
   |
note: required by a bound in `foo`
  --> $DIR/unstable-trait-suggestion.rs:9:11
   |
LL | fn foo<T: Unstable>(_: T) {}
   |           ^^^^^^^^ required by this bound in `foo`
```

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->


Fix #133511.

estebank committed:
Don't suggest restricting bound with unstable traits on stable

On nightly, we mention the trait is unstable

```
error[E0277]: the trait bound `T: Unstable` is not satisfied
  --> $DIR/unstable-trait-suggestion.rs:13:9
   |
LL |     foo(t)
   |     --- ^ the trait `Unstable` is not implemented for `T`
   |     |
   |     required by a bound introduced by this call
   |
note: required by a bound in `foo`
  --> $DIR/unstable-trait-suggestion.rs:9:11
   |
LL | fn foo<T: Unstable>(_: T) {}
   |           ^^^^^^^^ required by this bound in `foo`
help: consider restricting type parameter `T` but it is an `unstable` trait
   |
LL | pub fn demo<T: Unstable>(t: T) {
   |              ++++++++++
```

On stable, we don't suggest the trait at all

```
error[E0277]: the trait bound `T: Unstable` is not satisfied
  --> $DIR/unstable-trait-suggestion.rs:13:9
   |
LL |     foo(t)
   |     --- ^ the trait `Unstable` is not implemented for `T`
   |     |
   |     required by a bound introduced by this call
   |
note: required by a bound in `foo`
  --> $DIR/unstable-trait-suggestion.rs:9:11
   |
LL | fn foo<T: Unstable>(_: T) {}
   |           ^^^^^^^^ required by this bound in `foo`
```

compiler-errors commented on tests/ui/layout/rust-call-abi-not-a-tuple-ice-81974.stderr:
```diff
@@ -6,7 +6,7 @@ LL | impl<A, B> FnOnce<A> for CachedFun<A, B>
 note: required by a bound in `FnOnce`
   --> $SRC_DIR/core/src/ops/function.rs:LL:COL
-help: consider further restricting this bound
+help: consider further restricting this bound but it is an `unstable` trait
```
nit: `unstable` is not a keyword, and "but" feels awkward here... maybe put it in parens and use "although" instead?

> consider further restricting this bound (although this is an unstable trait)

compiler-errors replied:
maybe "although note that it is unstable" in the parentheses?

juntyr replied:
I also think the "it" here is a bit unclear, on first reading I thought "it" referred to `T` and was then thrown off by `T` being an unstable trait

Perhaps the trait could be named or something like "the new bound" or "the bound added below" be used instead?

compiler-errors replied:
the "it" is the trait, not really the bound

compiler-errors replied:
sure, but if you're going to use "but" please still put it in parentheses or at least add a comma, since it's still a bit awkward grammatically

cramertj replied:
Drive-by comment: perhaps it would be easier if the types were explicitly named? For example:

```
error[E0277]: functions with the "rust-call" ABI must take a single non-self tuple argument
  --> $DIR/rust-call-abi-not-a-tuple-ice-81974.rs:31:5
   |
LL |     extern "rust-call" fn call_once(mut self, a: A) -> Self::Output {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Tuple` is not implemented for `A`
   |
help: consider adding the unstable trait `Tuple` to the bounds of `A`
   |
LL |     A: Eq + Hash + Clone + std::marker::Tuple,
   |                          ++++++++++++++++++++
```

Then the stability or non-stability can be added via the phrase "consider adding the trait `Tuple` to the bounds of `A`" / "consider adding the unstable trait `Tuple` to the bounds of A`".

Alternatively, it could be suggested via a separate note:

```
error[E0277]: functions with the "rust-call" ABI must take a single non-self tuple argument
  --> $DIR/rust-call-abi-not-a-tuple-ice-81974.rs:31:5
   |
LL |     extern "rust-call" fn call_once(mut self, a: A) -> Self::Output {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Tuple` is not implemented for `A`
   |
help: consider adding the trait `Tuple` to the bounds of `A`
   |
LL |     A: Eq + Hash + Clone + std::marker::Tuple,
   |                          ++++++++++++++++++++
note: the trait `Tuple` is unstable
```

estebank committed:
reword trait bound suggestion message to include the bounds

compiler-errors commented on tests/rustdoc-ui/synthetic-auto-trait-impls/projections-in-super-trait-bound-unsatisfied.stderr:
```diff
@@ -4,7 +4,7 @@ error[E0277]: the trait bound `C: Bar<5>` is not satisfied
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Bar<5>` is not implemented for `C`
    |
-help: consider further restricting this bound
+help: consider further restricting this bound with trait `Bar<5>`
```
Could you change this to use `print_only_trait_name` rather than printing the binder and the args?

I first of all don't think mentioning the name is really that useful, since it's already there in the suggestion, but if we're going to mention it at all, it's probably only worthwhile to suggest only the trait name.

compiler-errors replied:
Rendering the binder and substs runs the risk of printing a really overly long trait ref here, since we may have large types in the args.

estebank replied:
We could only mention the trait for unsafe ones (as I originally intended in the PR, to keep the diff small).

The only issue against using the "name only" is when we have more than one bound being suggested, but we have to handle that anyways...

compiler-errors replied:
I think it's fine either way, was just mentioning that it feels a bit redundant. Name only seems like a good compromise.

estebank replied:
Addressed. It adds a bunch of logic, but it does look better.

estebank replied:
> was just mentioning that it feels a bit redundant

Yeah, but that was a review request though. I feel like it might be useful for the sake of suggestions being presented without context. I don't feel strongly one way or another.

estebank committed:
Add test for lack of suggestion in stable

This test will break when `Step` gets stabilized, but punt until then.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12074595570/job/33672882613) [(plain)](https://github.com/rust-lang/rust/commit/c05bacb73abfb11c457e0e78ca0767142ed2e7f8/checks/33672882613/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.797 Building wheels for collected packages: reuse
#13 2.798   Building wheel for reuse (pyproject.toml): started
#13 3.050   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.051   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.051   Stored in directory: /tmp/pip-ephem-wheel-cache-gx1g54x1/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.055 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.457 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.458 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.986 Collecting virtualenv
#13 3.986 Collecting virtualenv
#13 4.027   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.096      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 64.1 MB/s eta 0:00:00
#13 4.152 Collecting platformdirs<5,>=3.9.1
#13 4.156   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.175 Collecting distlib<1,>=0.3.7
#13 4.186      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 88.1 MB/s eta 0:00:00
#13 4.220 Collecting filelock<4,>=3.12.2
#13 4.224   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.224   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.306 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.491 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      317376 kB
DirectMap2M:     8071168 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 30.36s
##[endgroup]
fmt check
Diff in /checkout/tests/run-make/missing-unstable-trait-bound/rmake.rs:14:
         .input("missing-bound.rs")
         .run_fail()
         .assert_stderr_not_contains("help: consider restricting type parameter `T`")
-        .assert_stderr_contains(r#"
+        .assert_stderr_contains(
   = note: required for `std::ops::Range<T>` to implement `Iterator`
   = note: required for `std::ops::Range<T>` to implement `Iterator`
-  = note: required for `std::ops::Range<T>` to implement `IntoIterator`"#);
+  = note: required for `std::ops::Range<T>` to implement `IntoIterator`"#,
 }
 
 
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/run-make/native-link-modifier-whole-archive/native_lib_in_src.rs" "/checkout/tests/run-make/native-link-modifier-whole-archive/indirectly_linked.rs" "/checkout/tests/run-make/native-link-modifier-whole-archive/indirectly_linked_via_attr.rs" "/checkout/tests/run-make/native-link-modifier-whole-archive/directly_linked_test_minus_whole_archive.rs" "/checkout/tests/run-make/native-link-modifier-whole-archive/directly_linked_test_plus_whole_archive.rs" "/checkout/tests/run-make/native-link-modifier-whole-archive/directly_linked.rs" "/checkout/tests/run-make/native-link-modifier-whole-archive/rmake.rs" "/checkout/tests/run-make/native-link-modifier-whole-archive/rlib_with_cmdline_native_lib.rs" "/checkout/library/std/src/io/impls/tests.rs" "/checkout/tests/run-make/foreign-rust-exceptions/foo.rs" "/checkout/tests/run-make/foreign-rust-exceptions/bar.rs" "/checkout/tests/run-make/foreign-rust-exceptions/rmake.rs" "/checkout/tests/run-make/compressed-debuginfo/foo.rs" "/checkout/tests/run-make/compressed-debuginfo/rmake.rs" "/checkout/library/std/src/io/buffered/bufreader.rs" "/checkout/library/std/src/io/buffered/bufwriter.rs" "/checkout/library/std/src/io/buffered/linewritershim.rs" "/checkout/library/std/src/io/buffered/bufreader/buffer.rs" "/checkout/library/std/src/io/buffered/tests.rs" "/checkout/library/std/src/io/buffered/linewriter.rs" "/checkout/library/std/src/io/buffered/mod.rs" "/checkout/library/std/src/io/copy.rs" "/checkout/library/std/src/io/impls.rs" "/checkout/library/std/src/io/tests.rs" "/checkout/tests/run-make/extern-fn-generic/test.rs" "/checkout/tests/run-make/extern-fn-generic/rmake.rs" "/checkout/tests/run-make/extern-fn-generic/testcrate.rs" "/checkout/library/std/src/io/util/tests.rs" "/checkout/library/std/src/io/stdio/tests.rs" "/checkout/library/std/src/io/mod.rs" "/checkout/library/std/src/io/util.rs" "/checkout/library/std/src/f32.rs" "/checkout/tests/run-make/extern-fn-reachable/dylib.rs" "/checkout/tests/run-make/panic-abort-eh_frame/foo.rs" "/checkout/tests/run-make/panic-abort-eh_frame/rmake.rs" "/checkout/library/std/src/thread/current.rs" "/checkout/library/std/src/thread/local.rs" "/checkout/tests/run-make/incr-test-moved-file/main.rs" "/checkout/tests/run-make/incr-test-moved-file/rmake.rs" "/checkout/library/std/src/thread/local/tests.rs" "/checkout/library/std/src/thread/local/dynamic_tests.rs" "/checkout/library/std/src/thread/scoped.rs" "/checkout/library/std/src/thread/spawnhook.rs" "/checkout/library/std/src/thread/tests.rs" "/checkout/library/std/src/thread/mod.rs" "/checkout/tests/run-make/return-non-c-like-enum-from-c/rmake.rs" "/checkout/tests/run-make/return-non-c-like-enum-from-c/nonclike.rs" "/checkout/tests/run-make/no-input-file/rmake.rs" "/checkout/tests/run-make/missing-unstable-trait-bound/missing-bound.rs" "/checkout/tests/run-make/missing-unstable-trait-bound/rmake.rs" "/checkout/tests/run-make/mixing-formats/bar2.rs" "/checkout/tests/run-make/mixing-formats/foo.rs" "/checkout/tests/run-make/mixing-formats/bar1.rs" "/checkout/tests/run-make/mixing-formats/baz2.rs" "/checkout/tests/run-make/mixing-formats/rmake.rs" "/checkout/tests/run-make/mixing-formats/baz.rs" "/checkout/tests/run-make/doctests-keep-binaries-2024/rmake.rs" "/checkout/tests/run-make/doctests-keep-binaries-2024/t.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/main.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/rmake.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/stable.rs" "/checkout/tests/run-make/panic-impl-transitive/panic-impl-provider.rs" "/checkout/tests/run-make/panic-impl-transitive/panic-impl-consumer.rs" "/checkout/library/std/src/io/stdio.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Nov 28 19:44:32 UTC 2024
  network time: Thu, 28 Nov 2024 19:44:32 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

estebank committed:
Use trait name instead of full constraint in suggestion message

```
help: consider restricting type parameter `T` with traits `Copy` and `Trait`
   |
LL | fn duplicate_custom<T: Copy + Trait>(t: S<T>) -> (S<T>, S<T>) {
   |                      ++++++++++++++
```

```
help: consider restricting type parameter `V` with trait `Copy`
   |
LL | fn index<'a, K, V: std::marker::Copy>(map: &'a HashMap<K, V>, k: K) -> &'a V {
   |                  +++++++++++++++++++
```

estebank committed:
Mention type parameter in more cases and don't suggest ~const bound already there

estebank committed:
Account for `impl Trait` in "add bound" suggestion message

estebank committed:
fix rustdoc test

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12076118554/job/33677002312) [(plain)](https://github.com/rust-lang/rust/commit/442ec3d906af2f83577f067b5b46a10a26b68905/checks/33677002312/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 28.0s done
#17 DONE 35.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [ui] tests/ui/moves/use_of_moved_value_copy_suggestions.rs stdout ----

error: /checkout/tests/ui/moves/use_of_moved_value_copy_suggestions.rs:50: unexpected help message: '50:9: 50:9: consider further restricting type parameter `T` with traits `Copy` and `Trait`'

error: /checkout/tests/ui/moves/use_of_moved_value_copy_suggestions.rs:58: unexpected help message: '58:9: 58:9: consider further restricting type parameter `T` with traits `Copy` and `Trait`'

error: /checkout/tests/ui/moves/use_of_moved_value_copy_suggestions.rs:65: unexpected help message: '65:27: 65:27: consider further restricting type parameter `T` with traits `Copy` and `Trait`'
error: /checkout/tests/ui/moves/use_of_moved_value_copy_suggestions.rs:50: expected help message not found: consider further restricting this bound

error: /checkout/tests/ui/moves/use_of_moved_value_copy_suggestions.rs:58: expected help message not found: consider further restricting this bound


error: /checkout/tests/ui/moves/use_of_moved_value_copy_suggestions.rs:65: expected help message not found: consider further restricting this bound

error: 3 unexpected errors found, 3 expected errors not found
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/moves/use_of_moved_value_copy_suggestions.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/moves/use_of_moved_value_copy_suggestions" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- unexpected errors (from JSON output) ---
--- unexpected errors (from JSON output) ---
HELP MESSAGEline  50: 50:9: 50:9: consider further restricting type parameter `T` with traits `Copy` and `Trait`
HELP MESSAGEline  58: 58:9: 58:9: consider further restricting type parameter `T` with traits `Copy` and `Trait`
HELP MESSAGEline  65: 65:27: 65:27: consider further restricting type parameter `T` with traits `Copy` and `Trait`
--- not found errors (from test file) ---
HELP MESSAGEline  50: consider further restricting this bound
HELP MESSAGEline  58: consider further restricting this bound
HELP MESSAGEline  65: consider further restricting this bound
```

</details>

estebank committed:
Do not talk about "trait `<Foo = Bar>`"

Pass in an appropriate `Option<DefId>` in more cases from hir ty lowering.

compiler-errors commented:
I'll review later

estebank committed:
Use run-make `diff` output for stable output test

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 05e671451dcc1ef468e8e73777b2b7be92f81db2 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 05e671451dcc1ef468e8e73777b2b7be92f81db2 -->
<!-- homu: {"type":"Approved","sha":"05e671451dcc1ef468e8e73777b2b7be92f81db2","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

cramertj commented on tests/ui/associated-types/hr-associated-type-projection-1.stderr:
```diff
@@ -16,7 +16,7 @@ LL | trait UnsafeCopy<'a, T: Copy>
 LL |     for<'b> <Self as UnsafeCopy<'b, T>>::Item: std::ops::Deref<Target = T>,
    |                                                                ^^^^^^^^^^ required by this bound in `UnsafeCopy`
-help: consider further restricting this bound
+help: consider further restricting this bound with  `<Target = T>`
```
nit: It looks like an extra space snuck in here.

estebank replied:
That output is already not great... I think I'll do a iollow up just reverting to the previous message for that one case. There's another for "with `, Output = Ty`" which I don't love either.

estebank committed:
Tweak wording

estebank commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 25ad0478cb0975f10db6db3abb28d54df8e07430 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 25ad0478cb0975f10db6db3abb28d54df8e07430 -->
<!-- homu: {"type":"Approved","sha":"25ad0478cb0975f10db6db3abb28d54df8e07430","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
actually conflict-prone it seems

@bors rollup=iffy

jieyouxu commented:
@bors p=1 (conflict-prone)

bors commented:
:hourglass: Testing commit 25ad0478cb0975f10db6db3abb28d54df8e07430 with merge f415c07494b98e4559e4b13a9c5f867b0e6b2444...
<!-- homu: {"type":"BuildStarted","head_sha":"25ad0478cb0975f10db6db3abb28d54df8e07430","merge_sha":"f415c07494b98e4559e4b13a9c5f867b0e6b2444"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12220529173/job/34090737013)
Approved by: compiler-errors
Pushing f415c07494b98e4559e4b13a9c5f867b0e6b2444 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12220529173/job/34090737013"},"merge_sha":"f415c07494b98e4559e4b13a9c5f867b0e6b2444"} -->

rust-timer commented:
Finished benchmarking commit (f415c07494b98e4559e4b13a9c5f867b0e6b2444): [comparison URL](https://perf.rust-lang.org/compare.html?start=728f2daab42ba8f1b3d5caab62495798d1eabfa1&end=f415c07494b98e4559e4b13a9c5f867b0e6b2444&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=728f2daab42ba8f1b3d5caab62495798d1eabfa1&end=f415c07494b98e4559e4b13a9c5f867b0e6b2444&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.3% | [2.3%, 2.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=728f2daab42ba8f1b3d5caab62495798d1eabfa1&end=f415c07494b98e4559e4b13a9c5f867b0e6b2444&stat=max-rss)
<details>
<summary>Results (primary 0.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.8%, 2.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.4%, -1.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.7%  | [-1.4%, 2.8%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=728f2daab42ba8f1b3d5caab62495798d1eabfa1&end=f415c07494b98e4559e4b13a9c5f867b0e6b2444&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=728f2daab42ba8f1b3d5caab62495798d1eabfa1&end=f415c07494b98e4559e4b13a9c5f867b0e6b2444&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 766.806s -> 766.757s (-0.01%)
**Artifact size**: 330.83 MiB -> 330.86 MiB (0.01%)
